### PR TITLE
feat: use `leantar` in `lake exe cache`

### PR DESCRIPTION
**Pull request by digama0**
This hooks up the new `leantar` tool from https://github.com/digama0/leangz to `lake exe cache`. It uses an olean compressor to achieve much smaller file sizes, and faster decompression. See the [Zulip discussion](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/leangz.3A.20olean.20file.20compressor/near/369738648).

The new files have a new file extension, `123.ltar` instead of `123.tar.gz`. This implementation does not multiplex between them, so it will be a hard reset. In fact, the new cache doesn't even know how to clear the old cache files, so you might need to do that manually. Switching between branches will still work during the interim, since the cache will have both kinds of file and use the right ones.

---

- [x] depends on: #5409

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Cache/IO.lean [hunk 1]
```diff
@@ -136,6 +148,38 @@ def validateCurl : IO Bool := do
...
+def validateLeanTar : IO Unit := do
+  if (← LEANTARBIN.pathExists) then return
+  if let some version ← some <$> runCmd "leantar" #["--version"] <|> pure none then
+    let "leantar" :: v :: _ := version.splitOn " "
+      | throw $ IO.userError "Invalidly formatted response from `leantar --version`"
+    let maj :: min :: patch :: _ := v.splitOn "."
+      | throw $ IO.userError "Invalidly formatted version of `leantar`"
+    let version := (maj.toNat!, min.toNat!, patch.toNat!)
+    -- let _ := @lexOrd
+    -- let _ := @leOfOrd
```
Comment (semorrison) [inline Cache/IO.lean:160]:
```suggestion
```

Comment (digama0) [reply Cache/IO.lean]: these are here on purpose, to be uncommented in the event that the `=` becomes a `>=`. But I've since refactored this part anyway.

**Comment by semorrison** [comment]
All working for me now, and a massive speedup:
```
# Before #5710
lake clean && lake exe cache get!   # 343s
lake exe cache get                  # 17s

# After #5710
lake clean && lake exe cache get!   # 49s
lake exe cache get                  # 4s
```

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#5409
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by digama0** [comment]
Oh wow, I just had the pleasure of running `lake exe cache get` on a low bandwidth hotel wifi, and it took 25 minutes. A (projected) 7x decrease to 3.57 minutes would make quite a big difference!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#5409~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
bors merge

File: Cache/IO.lean [hunk 2]
```diff
@@ -51,12 +59,23 @@ def CURLBIN :=
   -- change file name if we ever need a more recent version to trigger re-download
   IO.CACHEDIR / s!"curl-{CURLVERSION}"
 
+/-- leantar version at https://github.com/digama0/leangz -/
+def LEANTARVERSION :=
+  "0.1.2"
```
Comment (gebner) [inline Cache/IO.lean:64]: This PR should wait until we have statically linked builds of leantar.

**Review by gebner — DISMISSED** [review]

File: Cache/IO.lean (hunk 2 above)
Comment (digama0) [reply Cache/IO.lean]:
```
$ ldd ~/.cache/mathlib/leantar-0.1.3
        statically linked
```

File: Cache/IO.lean [hunk 3]
```diff
@@ -136,6 +156,45 @@ def validateCurl : IO Bool := do
...
+      if System.Platform.getIsOSX () then s!"{arch}-apple-darwin"
+      else s!"{arch}-unknown-linux-musl"
+  IO.println s!"leantar is too old; downloading more recent version"
+  IO.FS.createDirAll IO.CACHEDIR
+  let ext := if win then "zip" else "tar.gz"
+  let _ ← runCmd "curl" #[
+    s!"https://github.com/digama0/leangz/releases/download/v{LEANTARVERSION}/leantar-v{LEANTARVERSION}-{target}.{ext}",
+    "-L", "-o", s!"{LEANTARBIN}.{ext}"]
+  let _ ← runCmd "tar" #["-xf", s!"{LEANTARBIN}.{ext}",
+    "-C", IO.CACHEDIR.toString, "--strip-components=1"]
```
Comment (gebner) [inline Cache/IO.lean:195]: Does this work on windows (i.e., with zip)?

**Comment by gebner** [comment]
I am happy with the static linking issue now.  The branch works fine on nixos.

@fpvandoorn complained about windows performance on zulip.  Is this still an issue (is it even a regression) and is there something we can do about it?

**Comment by digama0** [comment]
We investigated it and it doesn't seem like there is much we can do about it, the bottleneck is the write itself and I think windows is just slow at writing lots of files. I don't think it is a regression compared to the original `tar` script, but perhaps @fpvandoorn could do a clean test and confirm this.

File: Cache/IO.lean (hunk 3 above)
Comment (digama0) [reply Cache/IO.lean:195]: It seems to work... not sure how universally available `tar` is on windows but this exact command was tested and works on at least a few machines.

**Comment by fpvandoorn** [comment]
When testing a bit more carefully yesterday, I think now that `leantar` is still quite a bit quicker than the original `tar`. I don't have time to test it atm, but I'm happy if this gets merged.

File: Cache/IO.lean (hunk 3 above)
Comment (gebner) [reply Cache/IO.lean:195]: :scream: 

> The old cache was also using tar for unpacking on windows and I don't think we have received any complaints about it.

The old cache used tar to unpack tar files.  This line here uses tar to unpack zip files.  But if it works. :shrug:

**Comment by semorrison** [comment]
bors seems to not be working. I'll try again.

**Comment by semorrison** [comment]
bors merge

**Comment by semorrison** [comment]
I think bors is failing here because it has already merged this PR once (my fault, I merged too early and had to revert). I'm just going to squash merge now.

**Pull request merged**